src/csharp/Tests/UnitTests/LoggingTests.cs

```diff
@@ -0,0 +1,199 @@
+            {
+                _ = Assert.Throws(typeof(InvalidOperationException), () =>
+                {
+                    using (LoggingTracer trace = new LoggingTracer())
+                    {
+                    }
+                });
+
+                Assert.AreEqual(0, Device.GetInstalledCount());
+                Assert.AreEqual(1, tracer.LogMessages.Count);
+                Assert.IsTrue(tracer.LogMessages[0].EndsWith("[Warning] File@1234: Message", StringComparison.Ordinal));
+            }
+        }
+
+        /// <summary>
+        /// Tests that verify that multiple threads can create logging tracers and that they get
+        /// only the calls for their thread and that the lifetime of the tracer on other threads
+        /// don't impact other threads.
+        /// </summary>
+        [Test]
+        public void LoggingTracerMultiThreaded()
+        {
+            this.SetMessageHandlerImplementation();
+
+            Task threadA = Task.Run(() => ThreadProc(1, 50));
+            Task threadB = Task.Run(() => ThreadProc(10, 10));
+            Task threadC = Task.Run(() => ThreadProc(11, 9));
+
+            threadA.Wait();
+            threadB.Wait();
+            threadC.Wait();
+        }
+
+        private static void ThreadProc(int delay, int count)
+        {
+            // Make sure that there are calls and times that a thread exists where the tracer isn't active.
+            Assert.AreEqual(0, Device.GetInstalledCount());
+            Thread.Sleep(delay);
+
+            using (LoggingTracer tracer = new LoggingTracer())
+            {
+                for (int i = 0; i < count; ++i)
+                {
+                    Assert.AreEqual(0, Device.GetInstalledCount());
+                    Thread.Sleep(delay);
+                }
+
+                Assert.AreEqual(count, tracer.LogMessages.Count);
+            }
+
+            Assert.AreEqual(0, Device.GetInstalledCount());
+        }
+
+        /// <summary>
+        /// This sets up an implementation for the set debug message handler and for get installed
+        /// count to enable testing. Every call to get the installed count should result in one
+        /// message being generated.
+        /// </summary>
+        private void SetMessageHandlerImplementation()
+        {
+            this.nativeK4a.SetImplementation(@"
+
+k4a_logging_message_cb_t *g_message_cb;
+
+uint32_t k4a_device_get_installed_count(void)
+{
+    if(g_message_cb != NULL)
+    {
+        g_message_cb(NULL, K4A_LOG_LEVEL_WARNING, ""File"", 1234, ""Message"");
+    }
+
+    return 0;
+}
+
+k4a_result_t k4a_set_debug_message_handler(
+    k4a_logging_message_cb_t *message_cb,
+    void *message_cb_context,
+    k4a_log_level_t min_level)
+{
+    STUB_ASSERT(message_cb != NULL);
+    STUB_ASSERT(message_cb_context == NULL);
+    STUB_ASSERT(min_level >= K4A_LOG_LEVEL_WARNING);
+
+    g_message_cb = message_cb;
+
+    return K4A_RESULT_SUCCEEDED;
+}");
```

Comment:
@Brent-A For some reason this isn't getting picked up if I run all of the tests at the same time. It is using the k4a_set_debug_message_handler that was already defined. Do you know why?